util/src/main/java/org/folio/reservoir/util/readstream/XmlFixerMapper.java

```diff
@@ -42,26 +44,76 @@ static boolean is4byteSequence(byte b) {
     return (b & 248) == 240;
   }
 
-  void handleSequence(Buffer input, byte leadingByte) {
+  boolean handleSequence(Buffer input, byte leadingByte) {
     // the order of checks doesn't matter here but the most occurring
     // check is listed first. There will always be more continuation bytes
     // than leading bytes.
     if (isContinuation(leadingByte)) {
-      if (sequenceLength > 0) {
-        sequenceLength--;
-      } else {
+      if (sequenceLength == 0) {
         skipByte(input);
+        return false;
       }
-    } else if (is2byteSequence(leadingByte)) {
+      sequenceLength--;
+      if (sequenceLength == 0) {
+        flushSequence(input);
+      }
+      return false;
+    }
+    checkSkipSequence(input);
+    if (!ended && front + 3 + ASCII_LOOKAHED >= input.length()) {
+      incomplete(input, input.length());
+      return true;
+    }
+    if (is2byteSequence(leadingByte)) {
       sequenceLength = 1;
     } else if (is3byteSequence(leadingByte)) {
       sequenceLength = 2;
     } else if (is4byteSequence(leadingByte)) {
       sequenceLength = 3;
     } else {
-      sequenceLength = 0;
       skipByte(input);
+      return false;
     }
+    start = front;
+    return false;
+  }
+
+  void checkSkipSequence(Buffer input) {
+    if (sequenceLength > 0) {
+      skipSequence(input);
+    }
+  }
+
+  void skipSequence(Buffer input) {
+    result.appendBuffer(input, tail, start - tail);
+    for (int i = start; i < front; i++) {
+      byte b = input.getByte(i);
+      if (isAscii(b)) {
+        result.appendByte(b);
+      }
+    }
+    tail = front;
+    start = -1;
+    sequenceLength = 0;
+    numberOfFixes++;
+  }
+
+  void flushSequence(Buffer input) {
```

Comment:
This is great. The only thing I wonder about if we should write `&#xFFFD;` into output for the sequences that cannot be recovered instead of skipping them?

Reply:
ok. At least we have a constant default. It would be trivial to make that "" and, thus, remove bad chars.